### PR TITLE
Inject controller config into the uniter

### PR DESCRIPTION
**Pull request by SimonRichardson**
This starts the injection of the controller config service into the unit agent. Unfortunately, the tests are a bit of a mess. So I've spent some time splitting the tests into individual files. This should hopefully make it easier to understand where a test should be.

Also, the mocks have been removed to prevent others from using other packages mocks. They're not your mocks, they're someone else's.

There is still some work to sort out the construction of the facade, but we can do that in another change.

In addition, I've spent a lot time sorting out the API server in terms of using the right service.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Test everything.

```sh
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
$ juju add-model default
$ juju deploy ubuntu -n 3
```

## Links

JUJU-4332

### CODE REVIEW

File: juju/testing/apiserver.go [hunk 1]
```diff
@@ -412,6 +398,34 @@ func (s *ApiServerSuite) TearDownTest(c *gc.C) {
...
+func (s *ApiServerSuite) SeedControllerCloud(c *gc.C, runner coredatabase.TxnRunner) {
+	err := InsertDummyCloudType(context.Background(), s.TxnRunner())
+	c.Assert(err, jc.ErrorIsNil)
+	err = cloudbootstrap.InsertInitialControllerCloud(DefaultCloud)(context.Background(), runner)
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+// AllowCloudType is responsible for applying the cloud type to
+// the given database.
+func AllowCloudType(ctx context.Context, db database.TxnRunner, version int, name string) error {
```
Comment (tlm) [inline juju/testing/apiserver.go:417]: Should we move this to the cloud state? It's something that we may or may not need some where download the line?

File: juju/testing/apiserver.go [hunk 2]
```diff
@@ -555,6 +569,48 @@ func (s *ApiServerSuite) tearDownConn(c *gc.C) {
 	}
 }
 
+func (s *ApiServerSuite) SeedCAASCloud(c *gc.C) {
```
Comment (tlm) [inline juju/testing/apiserver.go:572]: Is this better off being in a CAAS test suite?

**Review by tlm — COMMENTED** [review]
haven't gone through everything yet. But so far lgtm. Just a few comments question.

File: juju/testing/apiserver.go (hunk 2 above)
Comment (SimonRichardson) [reply juju/testing/apiserver.go:572]: So, maybe we should have a CAASApiServerSuite, which only offers these methods and doesn't pollute the ApiServerSuite? Then we can just seed the CAAS cloud automatically.

**Review by manadart — APPROVED** [review]
It's good, just consider the suite organisation.
We can come around on some of the other stuff later.

File: juju/testing/apiserver.go (hunk 2 above)
Comment (SimonRichardson) [reply juju/testing/apiserver.go:572]: @tlm I'm going to sweep back around to this in a new PR. Easier to review.

File: juju/testing/apiserver.go [hunk 3]
```diff
@@ -555,6 +569,48 @@ func (s *ApiServerSuite) tearDownConn(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+	err = s.TxnRunner().Txn(context.Background(), func(ctx context.Context, tx *sqlair.TX) error {
+		return credentialstate.CreateCredential(ctx, tx, credUUID.String(), "dummy-credential", "caascloud", "admin", cred)
+	})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+// SeedDatabase the database with test values.
+func SeedDatabase(c *gc.C, runner coredatabase.TxnRunner, controllerConfig controller.Config) {
+	err := controllerconfigbootstrap.InsertInitialControllerConfig(controllerConfig)(context.Background(), runner)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply juju/testing/apiserver.go:590]: I'll revisit this one when I tackle the CAASAPIServerSuite.

File: cmd/jujud/agent/bootstrap_test.go [hunk 4]
```diff
@@ -119,7 +118,7 @@ func (s *BootstrapSuite) SetUpTest(c *gc.C) {
 	})
 	s.PatchValue(&corebase.UbuntuDistroInfo, "/path/notexists")
 	s.PatchValue(&extraBootstrapOpts, []database.BootstrapOpt{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud/agent/bootstrap_test.go:120]: Agreed. I'll have a look at undoing this work.

File: juju/testing/apiserver.go (hunk 2 above)
Comment (tlm) [reply juju/testing/apiserver.go:563]: Yeah that's fine. I have some ideas on it.

**Comment by tlm** [comment]
Thanks @SimonRichardson for this.

**Pull request merged**